packages/core/src/libraries/hook.ts

```diff
@@ -0,0 +1,72 @@
+import { Event, userInfoSelectFields } from '@logto/schemas';
+import { HookEventPayload, HookEvent } from '@logto/schemas/models';
+import { trySafe } from '@logto/shared';
+import { conditional, pick } from '@silverhand/essentials';
+import { got } from 'got';
+import type { Provider } from 'oidc-provider';
+
+import modelRouters from '#src/model-routers/index.js';
+import { findApplicationById } from '#src/queries/application.js';
+import { findUserById } from '#src/queries/user.js';
+import { getInteractionStorage } from '#src/routes/interaction/utils/interaction.js';
+
+const eventToHook: Record<Event, HookEvent> = {
+  [Event.Register]: HookEvent.PostRegister,
+  [Event.SignIn]: HookEvent.PostSignIn,
+  [Event.ForgotPassword]: HookEvent.PostResetPassword,
+};
+
+export type Interaction = Awaited<ReturnType<Provider['interactionDetails']>>;
+
+export const triggerInteractionHooksIfNeeded = async (
+  details?: Interaction,
+  userAgent?: string
+) => {
+  const userId = details?.result?.login?.accountId;
+  const sessionId = details?.jti;
+  const applicationId = details?.params.client_id;
+
+  if (!userId) {
+    return;
+  }
+
+  const interactionPayload = getInteractionStorage(details.result);
+  const { event } = interactionPayload;
+
+  const hookEvent = eventToHook[event];
+  const { rows } = await modelRouters.hook.client.readAll();
```

Comment:
Why not add a "where" condition here?

Reply:
you are right, actually we can

Reply:
oh it doesn't have the built-in support yet, add LOG-4928